cvmfs/receiver/payload_processor.cc

```diff
@@ -0,0 +1,118 @@
+
+PayloadProcessor::PayloadProcessor() : current_repo_(), num_errors_(0) {}
+
+PayloadProcessor::~PayloadProcessor() {}
+
+PayloadProcessor::Result PayloadProcessor::Process(
+    int fdin, const std::string& digest_base64, const std::string& path,
+    uint64_t header_size) {
+  const size_t first_slash_idx = path.find('/', 0);
+
+  current_repo_ = path.substr(0, first_slash_idx);
+
+  std::string header_digest;
+  if (!Debase64(digest_base64, &header_digest)) {
+    return kOtherError;
+  }
+
+  // Set up object pack deserialization
+  shash::Any digest = shash::MkFromHexPtr(shash::HexPtr(header_digest));
+  ObjectPackConsumer deserializer(digest, header_size);
+  deserializer.RegisterListener(&PayloadProcessor::ConsumerEventCallback, this);
+
+  int nb = 0;
+  ObjectPackBuild::State consumer_state = ObjectPackBuild::kStateContinue;
+  do {
+    std::vector<unsigned char> buffer(4096, 0);
+
+    nb = read(fdin, &buffer[0], buffer.size());
+    consumer_state = deserializer.ConsumeNext(nb, &buffer[0]);
+    if (consumer_state != ObjectPackBuild::kStateContinue &&
+        consumer_state != ObjectPackBuild::kStateDone) {
+      LogCvmfs(kLogCvmfs, kLogStderr,
+               "Error %d encountered when consuming object pack.",
+               consumer_state);
+      break;
+    }
+  } while (nb > 0 && consumer_state != ObjectPackBuild::kStateDone);
+
+  if (GetNumErrors() > 0) {
+    return kOtherError;
+  }
+
+  return kSuccess;
+}
+
+void PayloadProcessor::ConsumerEventCallback(
+    const ObjectPackBuild::Event& event) {
+  std::string path("");
+  if (event.object_type == ObjectPack::kCas) {
+    path = event.id.MakePath();
+  } else if (event.object_type == ObjectPack::kNamed) {
+    path = event.object_name;
+  } else {
+    // kEmpty - this is an error.
+    LogCvmfs(kLogCvmfs, kLogStderr, "Event received with unknown object.");
+    num_errors_++;
+    return;
+  }
+
+  const std::string hash_string = event.id.ToString(true);
+  const char suffix = hash_string[hash_string.size() - 1];
+
+  if (suffix == 'C') {
+    // Catalog
+    // Extract path names which are updated in this change set and return the
+    // list to the caller, to be used in a later call to rebuild catalogs
+  } else {
+    // Normal file
+
+    // Create a temporary path
+    const std::string tmp_path = CreateTempPath("/tmp/object_packs", 0666);
+    if (tmp_path.empty()) {
+      LogCvmfs(kLogCvmfs, kLogStderr, "Unable to create temporary path.");
+      num_errors_++;
+      return;
+    }
+
+    int fdout = open(tmp_path.c_str(), O_CREAT | O_WRONLY | O_TRUNC, 0600);
+    int nb = WriteFile(fdout, event.buf, event.buf_size);
+    if (static_cast<unsigned int>(nb) != event.buf_size) {
+      LogCvmfs(kLogCvmfs, kLogStderr, "Unable to write %s", tmp_path.c_str());
+      num_errors_++;
+      RemoveTree(tmp_path);
+      return;
+    }
+
+    // Atomically move to final destination
+    const std::string dest = "/srv/cvmfs/" + current_repo_ + "/data/" + path;
```

Comment:
We could add a ToDo here: I think at some point we'd like to use the spooler at this place.  This would allow us to commit objects to an S3 backend, or to a non-standard location in the file system.